### ISSUE TITLE
Make resolve results injectable in constructor

### ISSUE THREAD

**Issue by robert-bor** [angular/angular#10838]
```
[ ] bug report => search github for a similar issue or PR before submitting
[X] feature request
[ ] support request
```

First off, many thanks for your really great work on Angular. We love where this is going. Thanks! 

**Current behavior**

Given a component:

``` typescript
@Component({
    template: require('./book-detail.html')
})
export class BookDetail implements OnInit {

    private book : Book;

    constructor(private route: ActivatedRoute) {}

    ngOnInit(): void {
        this.route.data.subscribe((data: { book: Book }) => {
            this.book = data.book;
        });
    }

}
```

And a resolver:

``` typescript
@Injectable()
export class BookDetailResolve implements Resolve<any> {
    constructor(private bookService: BookService) {}

    resolve(route: ActivatedRouteSnapshot): Promise<Book> {
        let id : number = _.toNumber(route.params['id']);
        return this.bookService.getBook(id);
    }
}
```

And a configuration in the router: 

``` typescript
            path: ':id',
            component: BookDetail,
            resolve: {
              book: BookDetailResolve
            }
```

The way a book is passed to BookDetail is now done by having BookDetail implement OnInit and using the ngInit to read the book from the ActivatedRoute.

**Expected/desired behavior**

Ideally, it would be possible to have the following code:

``` typescript
@Component({
    template: require('./book-detail.html')
})
export class BookDetail {

    constructor(private book: Book) {}

}
```

In other words, the book as described in the route (and returned by BookDetailResolve) is injected directly into the constructor of BookDetail. This removes code which basically serves no purpose since resolve guarantees that all arguments are resolved prior to initializing the component. Code becomes predictable and the programming model is much easier to understand.

**What is the motivation / use case for changing the behavior?**

To reduce code which serves only as boilerplate and strive towards a cleaner and easier-to-understand codebase.
- **Angular version:** 2.0.0-rc.5

**Comment by DzmitryShylovich** [comment]
Very limited feature as to me. For example, u navigate from `post/1` => `post/2`. Component will be reused, but resolver will fetch new data and what will happen next? We can't inject new data into already created component.

**Comment by robert-bor** [comment]
Whether the feature is limited or not depends on whether you argue from component reuse. If you do, then you are right and the feature is indeed limited. But perhaps there is a way to see the bigger picture. Personally, I think component reuse in Angular is overrated. More so when it hampers the simplicity of the programming model. Is it really so objectionable to recreate the component for every screen?

**Comment by Colorfulstan** [comment]
Not sure if im missing the point here, so excuse me if misunderstood the problem. 
Since I used ui-router for angularJS already I'm using ui-router for ng2 too and didn't really bother learning much about the original ng2 router so far.

With ui-router you define your resolves within a state, which has a component associated to that state, that will be created when transitioning to the state.

You then can use the resolves as Input within your component.
Using Typescript this might look something like this then:
```
export const BOOK_STATE: Ng2StateDeclaration = {
    name     : 'books.detail',
    url      : '/:id',
    component: BookDetail,
    resolve  : [
        {
            provide   : 'book',
            useFactory: function (bookService:BookService, bookId:number) {
                return bookService.getBook(bookId)
            }, deps   : [BookService, 'id']
        }
    ]
}
```
```
@Component({
    template: require('./book-detail.html')
})
export class BookDetail {

    @Input() book : Book;
    constructor() {}
}
```

This allows to either input the Book as a resolve or as an attribute binding within the custom html tag, depending on which behaviour is needed (for example either as a Detail-page with it's own state, or a list of all book details).

**Comment by DzmitryShylovich** [comment]
Sounds like https://github.com/angular/angular/issues/9787
